File: services/qualification_service.go
package services

import (
	"context"
	"errors"

	"doctor-service/dto"
	"doctor-service/entities"
	"doctor-service/repositories"

	"github.com/google/uuid"
)

type QualificationService interface {
	AddQualification(ctx context.Context, doctorID uuid.UUID, req dto.AddQualificationRequest) (*dto.QualificationResponse, error)
	GetQualifications(ctx context.Context, doctorID uuid.UUID) ([]dto.QualificationResponse, error)
	DeleteQualification(ctx context.Context, qualificationID, doctorID uuid.UUID) error
}

type DefaultQualificationService struct {
	doctorRepo repositories.DoctorRepository
	qualRepo   repositories.QualificationRepository
}

func NewQualificationService(doctorRepo repositories.DoctorRepository, qualRepo repositories.QualificationRepository) QualificationService {
	return &DefaultQualificationService{doctorRepo: doctorRepo, qualRepo: qualRepo}
}

func (s *DefaultQualificationService) AddQualification(ctx context.Context, doctorID uuid.UUID, req dto.AddQualificationRequest) (*dto.QualificationResponse, error) {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorID)
	if err != nil || doc == nil {
		return nil, errors.New("doctor not found")
	}

	qual := &entities.DoctorQualification{
		DoctorID:       doc.ID,
		Degree:         req.Degree,
		Specialization: req.Specialization,
		College:        req.College,
		University:     req.University,
		YearCompleted:  req.YearCompleted,
	}

	if err := s.qualRepo.Create(ctx, qual); err != nil {
		return nil, err
	}

	return &dto.QualificationResponse{
		QualificationID: qual.QualificationID,
		Degree:          qual.Degree,
		Specialization:  qual.Specialization,
		College:         qual.College,
		University:      qual.University,
		YearCompleted:   qual.YearCompleted,
		CreatedAt:       qual.CreatedAt,
	}, nil
}

func (s *DefaultQualificationService) GetQualifications(ctx context.Context, doctorID uuid.UUID) ([]dto.QualificationResponse, error) {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorID)
	if err != nil || doc == nil {
		return nil, errors.New("doctor not found")
	}

	quals, err := s.qualRepo.FindByDoctorID(ctx, doc.ID)
	if err != nil {
		return nil, err
	}

	var res []dto.QualificationResponse
	for _, q := range quals {
		res = append(res, dto.QualificationResponse{
			QualificationID: q.QualificationID,
			Degree:          q.Degree,
			Specialization:  q.Specialization,
			College:         q.College,
			University:      q.University,
			YearCompleted:   q.YearCompleted,
			CreatedAt:       q.CreatedAt,
		})
	}
	return res, nil
}

func (s *DefaultQualificationService) DeleteQualification(ctx context.Context, qualificationID, doctorID uuid.UUID) error {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorID)
	if err != nil || doc == nil {
		return errors.New("doctor not found")
	}
	return s.qualRepo.Delete(ctx, qualificationID, doc.ID)
}

File: python_backend/services/qualification_service.py
from datetime import datetime
from entities.qualification import DoctorQualification


class QualificationService:
    def AddQualification(self, ctx, doctor_id, req):
        pass

    def GetQualifications(self, ctx, doctor_id):
        pass

    def DeleteQualification(self, ctx, qualification_id, doctor_id):
        pass


class DefaultQualificationService(QualificationService):
    def __init__(self, doctor_repo, qual_repo):
        self.doctorRepo = doctor_repo
        self.qualRepo = qual_repo

    def _to_response(self, qual):
        class QualificationResponse:
            pass

        resp = QualificationResponse()
        resp.qualification_id = str(qual.qualification_id)
        resp.degree = qual.degree
        resp.specialization = qual.specialization or ""
        resp.college = qual.college
        resp.university = qual.university
        resp.year_completed = qual.year_completed
        resp.created_at = qual.created_at or datetime.utcnow()
        return resp

    def AddQualification(self, ctx, doctor_id, req):
        doc = self.doctorRepo.FindByPublicID(str(doctor_id))
        if not doc:
            raise Exception("doctor not found")

        qual = DoctorQualification(
            doctor_id=str(doc.id),
            degree=req.degree,
            specialization=getattr(req, "specialization", None) or "",
            college=req.college,
            university=req.university,
            year_completed=req.year_completed,
        )
        self.qualRepo.Create(qual)
        return self._to_response(qual)

    def GetQualifications(self, ctx, doctor_id):
        doc = self.doctorRepo.FindByPublicID(str(doctor_id))
        if not doc:
            raise Exception("doctor not found")
        quals = self.qualRepo.FindByDoctorID(doc.id)
        return [self._to_response(q) for q in quals]

    def DeleteQualification(self, ctx, qualification_id, doctor_id):
        doc = self.doctorRepo.FindByPublicID(str(doctor_id))
        if not doc:
            raise Exception("doctor not found")
        return self.qualRepo.Delete(str(qualification_id), doc.id)
